Prevent Windows Display Sleep

EliEron opened the pull request:
This PR prevents Windows from turning off the display while a game is running. 

I also looked into implementing this for Linux but determined it to be too complex, as it requires D-Bus communication, and it would have been hard for me to test since I don't currently have any Linux test environment setup.

The placement and use of the class is based on the WindowsMultimediaTimerResolution class, as that is also a Windows exclusive feature that calls native code. Windows itself keeps tracks of threads that are preventing the display from sleeping and resumes normal behavior when they exit, so it was not necessary to create a method to manually enable the display sleep.

I happened to noticed some minor errors in the GTK Glade file while I was looking at it (One wrong and one missing tooltip) which I decided I might as well fix.

I am pretty new to Git and code contributions in general so I apologize if I have made any mistakes, and I welcome any feedback.

This PR partially fixes issue #1829, I say partially because it only implements it for Windows, and not for Linux And macOS.

Thog commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -542,6 +542,10 @@ private void CreateGameWindow(HLE.Switch device)
                 _windowsMultimediaTimerResolution = new WindowsMultimediaTimerResolution(1);
+                if (ConfigurationState.Instance.System.PreventWindowsDisplaySleep)
+                {
+                    WindowsDisplaySleep.Prevent();
```
The previous execution state should be restored after exiting a emulation.

Thog reviewed (changes requested):
Mostly lgtm but state need to be restored when going back from emulation context.

Also, I'm not a fan of OS dependent settings like that, it could be nice investigating how to implement the same behaviour on other OSes if possible (otherwise, the setting need to be hidden for non Windows systems on the UI)

EliEron commented:
I did consider manually restoring the execution state while writing the PR, but in my testing it didn't seem to be necessary. Windows appeared to do so on it's own, after exiting sleep was reenabled without having to change anything. But just to be safe I have now added code to manually restore the execution state.

As for making the GUI option appear on Windows only I am entirely open to that, but have no idea how to code that, figuring out how to add the config option to the GUI was complex enough as it was, with the half a dozen places that had to be modified. 

Maybe it would be better to make it a config file only option? As I doubt a lot of people would want to turn it off in the first place.

AcK77 commented on Ryujinx.Common/System/WindowsDisplaySleep.cs:
```diff
@@ -0,0 +1,31 @@
+using System;
+using System.Runtime.InteropServices;
+
+
```
Remove extra empty line

AcK77 commented on Ryujinx.Common/System/WindowsDisplaySleep.cs:
```diff
@@ -0,0 +1,31 @@
+        public void Prevent()
+        {
+            SetThreadExecutionState(EXECUTION_STATE.ES_CONTINUOUS | EXECUTION_STATE.ES_SYSTEM_REQUIRED |
+                                    EXECUTION_STATE.ES_DISPLAY_REQUIRED);  
```
Remove extra empty line

AcK77 reviewed (changes requested):
If it was me I’d just disabled sleeping as default when a game is loaded in the emulation context and restore the previous sleeping when the emulation context is closed. I don’t think it needs any configuration field/settings.
Additionally to that, the WindowsDisplaySleeping should be renamed with a more generic name (DisplaySleeping ?). And the OS check needs to be done in this class directly in case someone want to implement other OS specific implementations of this.

EliEron commented:
Alright I have gone ahead and renamed the file and moved the OS check inside the methods. 

As for the config option I would like to get a consensus from you guys. Personally I wouldn't mind removing the config either, but the issue report this PR was created to resolve specifically asked for an option, and I do know that some really people dislike applications making these kinds of changes without offering any option to turn it off.

AcK77 commented:
Well, if you run a game on your PC, there is no sleeping until you close the game.
It’s same here, so there is no good point to let the system sleeping while you running a game under the emulator... I don’t see a situation where this could be useful!

EliEron commented:
Good point, and like I said I'm certainly not against it personally.  Alright I've gone ahead and removed the GUI changes. 

Hopefully that should resolve both your and Thog's main concerns with the code.

AcK77 commented on Ryujinx.Common/Configuration/ConfigurationFileFormat.cs:
```diff
@@ -152,7 +152,7 @@ public class ConfigurationFileFormat
         /// </summary>
         public bool EnableFsIntegrityChecks { get; set; }
-
+        
```
Remove extra spaces

AcK77 commented on Ryujinx.Common/Configuration/ConfigurationState.cs:
```diff
@@ -207,7 +207,7 @@ public class SystemSection
             /// </summary>
             public ReactiveObject<bool> EnableFsIntegrityChecks { get; private set; }
-
+            
```
Remove extra spaces

AcK77 commented on Ryujinx.Common/Configuration/ConfigurationState.cs:
```diff
@@ -225,16 +225,16 @@ public class SystemSection
-                FsGlobalAccessLogMode   = new ReactiveObject<int>();
-                AudioBackend            = new ReactiveObject<AudioBackend>();
-                IgnoreMissingServices   = new ReactiveObject<bool>();
+                Language                   = new ReactiveObject<Language>();
```
Remove extra spaces, aligment is done between `{ }` and not on the whole file

AcK77 commented on Ryujinx.Common/Configuration/ConfigurationState.cs:
```diff
@@ -769,59 +769,59 @@ public void Load(ConfigurationFileFormat configurationFileFormat, string configu
                 configurationFileUpdated = true;
             }
-
+            
```
Remove extra spaces

AcK77 commented on Ryujinx.Common/System/DisplaySleep.cs:
```diff
@@ -0,0 +1,35 @@
+
+        static public void Prevent()
+        {
+            if (Environment.OSVersion.Platform == PlatformID.Win32NT)
```
You should use `if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))`

AcK77 commented on Ryujinx.Common/System/DisplaySleep.cs:
```diff
@@ -0,0 +1,35 @@
+        
+        static public void Restore()
+        {
+            if (Environment.OSVersion.Platform == PlatformID.Win32NT)
```
You should use `if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))`

AcK77 commented on Ryujinx.Common/Configuration/ConfigurationState.cs:
```diff
@@ -390,37 +390,37 @@ public ConfigurationFileFormat ToFileFormat()
-                AudioBackend              = System.AudioBackend,
-                IgnoreMissingServices     = System.IgnoreMissingServices,
-                GuiColumns                = new GuiColumns
+                Version                    = ConfigurationFileFormat.CurrentVersion,
```
Remove extra spaces, aligment is done between `{ }` and not on the whole file

AcK77 commented on Ryujinx.Common/Configuration/ConfigurationState.cs:
```diff
@@ -453,52 +453,52 @@ public ConfigurationFileFormat ToFileFormat()
-            Ui.CustomThemePath.Value               = "";
-            Ui.StartFullscreen.Value               = false;
-            Hid.EnableKeyboard.Value               = false;
+            Graphics.ResScale.Value                 = 1;
```
Remove extra spaces, aligment is done between `{ }` and not on the whole file

AcK77 commented on Ryujinx.Common/Configuration/ConfigurationState.cs:
```diff
@@ -769,59 +769,59 @@ public void Load(ConfigurationFileFormat configurationFileFormat, string configu
-            Hid.EnableKeyboard.Value               = configurationFileFormat.EnableKeyboard;
-            Hid.Hotkeys.Value                      = configurationFileFormat.Hotkeys;
-            Hid.InputConfig.Value                  = inputConfig;
+            Graphics.ResScale.Value                 = configurationFileFormat.ResScale;
```
Remove extra spaces, aligment is done between `{ }` and not on the whole file

AcK77 commented on Ryujinx/Ui/SettingsWindow.cs:
```diff
@@ -383,36 +383,36 @@ private void SaveSettings()
-            ConfigurationState.Instance.Graphics.AspectRatio.Value             = Enum.Parse<AspectRatio>(_aspectRatio.ActiveId);
-            ConfigurationState.Instance.Graphics.ResScale.Value                = int.Parse(_resScaleCombo.ActiveId);
-            ConfigurationState.Instance.Graphics.ResScaleCustom.Value          = resScaleCustom;
+            ConfigurationState.Instance.Logger.EnableError.Value                = _errorLogToggle.Active;
```
Remove extra spaces, aligment is done between `{ }` and not on the whole file

EliEron commented:
Addressed all of the comments, hopefully it should be ready for merge now.

AcK77 commented:
Needs a rebase

EliEron commented:
Rebase has been completed, I also squashed all of the commits since the history was getting quite messy.

AcK77 reviewed (approved):
Lgtm!
